worlds/spelunker/Client.py

```diff
@@ -0,0 +1,207 @@
+        mod.__package__ = f"worlds.{mod.__package__}"
+        mod.__name__ = f"worlds.{mod.__name__}"
+        sys.modules[mod.__name__] = mod
+        importer.exec_module(mod)
+    elif not os.path.isdir(os.path.splitext(bh_apworld_path)[0]):
+        raise AssertionError("Could not import worlds._bizhawk")
+
+from NetUtils import ClientStatus
+import worlds._bizhawk as bizhawk
+from worlds._bizhawk.client import BizHawkClient
+import time
+
+if TYPE_CHECKING:
+    from worlds._bizhawk.context import BizHawkClientContext
+else:
+    BizHawkClientContext = object
+
+# Add .apwl suffix to bizhawk client
+from worlds.LauncherComponents import SuffixIdentifier, components
+
+for component in components:
+    if component.script_name == "BizHawkClient":
+        component.file_identifier = SuffixIdentifier(
+            *(*component.file_identifier.suffixes, ".apsplunker")
+        )
+        break
+
+EXPECTED_ROM_NAME = "SPELUNKERAP"
+
+
+class SpelunkerClient(BizHawkClient):
+    game = "Spelunker"
+    system = ("NES")
+    location_map = location_table
+    received_deathlinks = 0
+    deathlink_all_clear = False
+
+    def __init__(self) -> None:
+        super().__init__()
+
+    async def validate_rom(self, ctx: BizHawkClientContext) -> bool:
+        from CommonClient import logger
+
+        try:
+            # Check ROM name/patch version
+            rom_name_bytes = (
+                await bizhawk.read(ctx.bizhawk_ctx, [(0x7030, 11, "PRG ROM")])
+            )[0]
+            rom_name = bytes([byte for byte in rom_name_bytes if byte != 0]).decode(
+                "ascii"
+            )
+            if not rom_name.startswith(EXPECTED_ROM_NAME):
+                logger.info(
+                    "ERROR: Rom is not valid!"
+                )
+                return False
+        except UnicodeDecodeError:
+            return False
+        except bizhawk.RequestFailedError:
+            return False  # Should verify on the next pass
+
+        ctx.game = self.game
+        ctx.items_handling = 0b111
+
+        death_link = await bizhawk.read(ctx.bizhawk_ctx, [(0x7052, 1, "PRG ROM")])
+        if death_link:
+            await ctx.update_death_link(bool(death_link[0]))
+        return True
+
+    async def set_auth(self, ctx: BizHawkClientContext) -> None:
+        from CommonClient import logger
+
+        slot_name_length = await bizhawk.read(ctx.bizhawk_ctx, [(0x7040, 1, "PRG ROM")])
+        slot_name_bytes = await bizhawk.read(
+            ctx.bizhawk_ctx, [(0x7041, slot_name_length[0][0], "PRG ROM")]
+        )
+        ctx.auth = bytes([byte for byte in slot_name_bytes[0] if byte != 0]).decode(
+            "utf-8"
+        )
+
+    def on_package(self, ctx: "BizHawkClientContext", cmd: str, args: dict) -> None:
+        if cmd != "Bounced":
+            return
+        if "tags" not in args:
+            return
+        if "DeathLink" in args["tags"] and args["data"]["source"] != ctx.slot_info[ctx.slot].name:

```

Comment:
This means that two players playing on the same slot won't send deathlinks to each other. If you don't care about supporting that, don't worry about it.

Reply:
I'll have to think on this one.